### DISCUSSION TITLE
How to display the "particle.state" in Parcels 3.0.0 output

### DISCUSSION THREAD

**Discussion by Caiyuan6** [Parcels-code/Parcels#1580] [Q&A]
Hello, @erikvansebille. 
I am currently using Parcels version 3.0.0. However, I noticed that the output does not display the value of `particle.state`. I would like to know how I can view the value of  `particle.state`. Is there a specific method or setting that allows the value of  `particle.state` to be visible in the output results?
Any suggestions or advice you provide would be greatly appreciated.

**Comment by erikvansebille** [comment]
By default, `particle.state` is not in the zarr-output because its write-status is set to False. Have you tried changing the `write_status` of `particle.state`? E.g. using [`pset.set_variable_write_status('state', True)`](https://docs.oceanparcels.org/en/latest/reference/particles.html#parcels.particleset.ParticleSet.set_variable_write_status)

**Reply by Caiyuan6** [reply]
I modified the `state = Variable('state', dtype=np.int32, initial=StatusCode.Evaluate, to_write=True)` in **particle.py** in Parcels, and I was able to get the value of `particle.state` in the zarr output. I haven't tried using [pset.set_variable_write_status('state', True)](https://docs.oceanparcels.org/en/latest/reference/particles.html#parcels.particleset.ParticleSet.set_variable_write_status), but I will give it a try. 

In addition, 

1. I need to observe the changes in the value of `particle.state`. In the output, I can only see the changes from `StatusCode.Evaluate` to `StatusCode.delete`, but I never see `StatusCode.ErrorOutOfBounds`. Why is this happening?
2. Using `print` in this code causes the program to crash. How can I modify it to avoid this issue?
Here's the code for reference:

```
def CheckOutOfBounds(particle, fieldset, time):
    print(OutBound.astype(str))
    print(state.astype(str))
    OutBound = particle.state
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()
```

**Reply by erikvansebille** [reply]
Hi @Cyuan-hub, below answers to your two questions

> 1. I need to observe the changes in the value of particle.state. In the output, I can only see the changes from StatusCode.Evaluate to StatusCode.delete, but I never see StatusCode.ErrorOutOfBounds. Why is this happening?

That is because these error codes are 'captured' by the kernel at the end of the loop, and then processed. The loop can only complete with a `StatusCode.Evaluate` or `StatusCode.delete`, so you won't indeed see other StatusCodes in the output. A workaround would be to define your own `Variable` and set that to the value of `particle.state`. Something like

```python

MyParticle = JITParticle.add_variable(Variable('printstate'))

def PrintState(particle, fieldset, time):
    particle.printstate = particle.state

pset.execute([<YOUR OTHER KERNELS HERE>, PrintState], ...)
```

> 2. Using print in this code causes the program to crash. How can I modify it to avoid this issue?

The print function is (in JIT-mode) pretty limited in what it can do, because we have to hand-convert the print-statement from python to C++ [see this part of the code if you're interested](https://github.com/OceanParcels/parcels/blob/168e4045a2faace7999253d5342d733fc77ec059/parcels/compilation/codegenerator.py#L510-L539). See [the Kernels section of the parcels overview tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_parcels_structure.html#3.-Kernels) for some examples of what kind of print-statements _can_ be used in JIT-mode. 
Alternative, if you do want to use more complicated printing (especially for debugging), is to switch to Scipy-mode, by replacing `JITParticle` to `ScipyParticle` in your code.

**Reply by Caiyuan6** [reply]
Hi, @erikvansebille, thanks for your helpful response and guidance. Sorry for the delay in getting back to you; things have been pretty hectic lately. I'll keep you updated on any progress or if I run into any issues. Thaks a lot.
